tests/fermi/test_fermionic.py

```diff
@@ -168,3 +168,233 @@ def test_init_error(self, operator):
+                (3, 3): "-",
+                (4, 0): "+",
+                (5, 4): "-",
+            }
+        ): (-0 - 2j),
+        FermiWord(
+            {
+                (0, 0): "+",
+                (1, 3): "-",
+                (2, 0): "+",
+                (3, 4): "-",
+                (4, 0): "+",
+                (5, 1): "-",
+            }
+        ): -0.615,
+        FermiWord(
+            {
+                (0, 0): "+",
+                (1, 3): "-",
+                (2, 0): "+",
+                (3, 4): "-",
+                (4, 0): "+",
+                (5, 0): "-",
+            }
+        ): (-0 - 2j),
+        FermiWord(
+            {
+                (0, 0): "+",
+                (1, 3): "-",
+                (2, 0): "+",
+                (3, 4): "-",
+                (4, 0): "+",
+                (5, 3): "-",
+                (6, 0): "+",
+                (7, 4): "-",
+            }
+        ): 0.25,
+    }
+)
+
+
+class TestFermiSentence:
+    def test_missing(self):
+        """Test the result when a missing key is indexed."""
+        fw = FermiWord({(0, 0): "+", (1, 1): "-"})
+        new_fw = FermiWord({(0, 2): "+", (1, 3): "-"})
+        fs = FermiSentence({fw: 1.0})
+
+        assert new_fw not in fs.keys()
+        assert fs[new_fw] == 0.0
+
+    def test_set_items(self):
+        """Test that we can add to a FermiSentence."""
+        fw = FermiWord({(0, 0): "+", (1, 1): "-"})
+        fs = FermiSentence({fw: 1.0})
+
+        new_fw = FermiWord({(0, 2): "+", (1, 3): "-"})
+        assert new_fw not in fs.keys()
+
+        fs[new_fw] = 3.45
+        assert new_fw in fs.keys() and fs[new_fw] == 3.45
+
+    tup_fs_str = (
+        (fs1, "1.23 * '0+ 1-'\n" + "+ 4j * '0+ 0-'\n" + "+ -0.5 * '0+ 3- 0+ 4-'"),
+        (fs2, "-1.23 * '0+ 1-'\n" + "+ (-0-4j) * '0+ 0-'\n" + "+ 0.5 * '0+ 3- 0+ 4-'"),
+        (fs3, "-0.5 * '0+ 3- 0+ 4-'\n" + "+ 1 * ''"),
+        (fs4, "1 * ''"),
+        (fs5, "0 * ''"),
+    )
+
+    @pytest.mark.parametrize("fs, str_rep", tup_fs_str)
+    def test_str(self, fs, str_rep):
+        """Test the string representation of the FermiSentence."""
+        print(str(fs))
+        assert str(fs) == str_rep
+        assert repr(fs) == str_rep
+
+    tup_fs_wires = (
+        (fs1, {0, 1, 3, 4}),
+        (fs2, {0, 1, 3, 4}),
+        (fs3, {0, 3, 4}),
+        (fs4, set()),
+    )
+
+    @pytest.mark.parametrize("fs, wires", tup_fs_wires)
+    def test_wires(self, fs, wires):
+        """Test the correct wires are given for the FermiSentence."""
+        assert fs.wires == wires
+
+    @pytest.mark.parametrize("fs", (fs1, fs2, fs3, fs4))
+    def test_copy(self, fs):
+        """Test that the copy is identical to the original."""
+        copy_fs = copy(fs)
+        deep_copy_fs = deepcopy(fs)
+
+        assert copy_fs == fs
+        assert deep_copy_fs == fs
+        assert copy_fs is not fs
+        assert deep_copy_fs is not fs
+
+    tup_fs_mult = (  # computed by hand
```

Comment:
Can you add a test where there are cross terms that need to be collected. I don't think this case is covered. This could be something like: 

```
fs_1 = FermiSentence({fw2: 1, fw3: 1, fw4: 1})
fs_2 = FermiSentence({fw4: 1, fw2: 1})

fs_1 * fs_2
```
This way we have 2 terms which are `fw2` that then need to be collected together in the dict 👍🏼

Reply:
Good catch, this was not working - the second occurrence of `fw2` was replacing the first instead of being added on

Reply:
The initial implementation seems to be a fair bit faster than looping over to check for duplicates (even with the addition of a check that `len(set(keys))==len(keys)`, so I left it as is if there's no overlap, and then it deals with grouping as a special case. 

Lmk if that seems unnecessary or if you think of a more efficient way of dealing with grouping terms after multiplying!

Reply:
I think there is a more efficient way, lets chat about it after the core standup

Reply:
updated as discussed :)